### PR TITLE
Add collision and boundary schemes for PS

### PR DESCRIPTION
**Pull request by MikeW097**
The modifications are rebased on the current repository.

### CODE REVIEW

**Comment by haraldkl** [comment]
Is this a replacement for #12 then? Does this mean we should close #12?

**Review by haraldkl — APPROVED** [review]
Looks fine to me. Thanks a lot. Could we add the added testcases to the recheck?

**Comment by haraldkl** [comment]
I've added the `cylinder2d` setup to the recheck, and all tests pass:

```
Solver: musubi
Compiler: GFORTRAN in version 11.2.0
--------------------------------------------------------------------------------------------------------------------------------------
|   changeset    |  success  |   val_method    | result  |     time |          MLUPs | perfdiff(%) | Testcase                       |
--------------------------------------------------------------------------------------------------------------------------------------
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   1.92 s |            --- |         --- | gaussianPulse_musubi           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  18.10 s |            --- |         --- | C2D_Simple_BGK_musubi          |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  20.22 s |            --- |         --- | C2D_Simple_rBGK_musubi         |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   4.55 s |            --- |         --- | C2D_Simple_rrBGK_musubi        |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   7.16 s |            --- |         --- | C2D_Simple_prrBGK_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   7.25 s |            --- |         --- | C2D_Simple_hrrBGK_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   5.81 s |            --- |         --- | C2D_Simple_drtBGK_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  10.54 s |            --- |         --- | C2D_BC_MfrBB_PressExpol_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  19.83 s |            --- |         --- | C2D_BC_MfrEq_PressEq_musubi    |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   9.91 s |            --- |         --- | C2D_BC_PressExpol_PressExpol_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  27.19 s |            --- |         --- | C2D_BC_VelBB_PressExpol_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  47.13 s |            --- |         --- | C2D_BC_VelBFL_PressExpol_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  19.77 s |            --- |         --- | C2D_BC_VelEq_PressEq_musubi    |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   4.92 s |            --- |         --- | C2D_BC_VelNonEqExpol_PressNonEqExpol_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  57.33 s |            --- |         --- | C3D_Sph_ML_LES_D19_DRT_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  59.14 s |            --- |         --- | C3D_Sph_ML_LES_D19_HRR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  53.92 s |            --- |         --- | C3D_Sph_ML_LES_D19_PRR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  55.25 s |            --- |         --- | C3D_Sph_ML_LES_D19_R_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  57.91 s |            --- |         --- | C3D_Sph_ML_LES_D19_RR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  36.78 s |            --- |         --- | C3D_Sph_ML_LES_D27_CUM17_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  42.16 s |            --- |         --- | C3D_Sph_ML_LES_D27_DRT_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  41.91 s |            --- |         --- | C3D_Sph_ML_LES_D27_HRR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  41.36 s |            --- |         --- | C3D_Sph_ML_LES_D27_PRR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  41.24 s |            --- |         --- | C3D_Sph_ML_LES_D27_R_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  42.64 s |            --- |         --- | C3D_Sph_ML_LES_D27_RR_BGK_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  46.17 s |            --- |         --- | TC_SL_MuskerFP_musubi          |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  65.81 s |            --- |         --- | TC_SL_MusNewton_musubi         |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  59.83 s |            --- |         --- | TC_SL_ReichFP_musubi           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  89.92 s |            --- |         --- | TC_SL_PowerLaw_musubi          |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   0.98 s |            --- |         --- | ABS_Pulse_Radial_musubi        |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   1.17 s |            --- |         --- | ABS_Pulse_Plane_musubi         |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   0.79 s |            --- |         --- | ABS_Pulse_Box_musubi           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  64.25 s |            --- |         --- | ABS_Pulse_Box3D_musubi         |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  25.57 s |            --- |         --- | ABS_LineSrc2D_musubi           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  44.72 s |            --- |         --- | ABS_cyl2D_musubi               |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  45.61 s |            --- |         --- | TO_cyl2D_musubi                |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  45.11 s |            --- |         --- | VSL_Radial_musubi              |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   1.65 s |            --- |         --- | VSL_Box2D_musubi               |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  48.09 s |            --- |         --- | VSL_Box3D_musubi               |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  21.84 s |            --- |         --- | VSL_Plane_musubi               |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   1.94 s |            --- |         --- | gaussianPulse_incomp           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  26.25 s |            --- |         --- | TGV_Simple_Re800               |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  55.11 s |            --- |         --- | TGV_Simple_Re1600              |
|  2a225a631e22  |    OK     |   difference    | 0.000%  | 310.07 s |            --- |         --- | TGV_LES_WALE                   |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  30.42 s |            --- |         --- | TGV_LES_Vreman                 |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  28.69 s |            --- |         --- | TGV_LES_SmagPDF                |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  35.90 s |            --- |         --- | TGV_LES_SmagGradU              |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  23.38 s |            --- |         --- | C2D_Cyl_SL_Incomp_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  46.90 s |            --- |         --- | C2D_Cyl_ML_Incomp_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  15.86 s |            --- |         --- | C3D_Simple_Incomp_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  31.31 s |            --- |         --- | COC_CouetteFlow_Incomp_musubi  |
|  2a225a631e22  |    OK     |   difference    | 0.000%  | 159.40 s |            --- |         --- | COC_CouetteFlow_ML_Incomp_musubi |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  22.38 s |            --- |         --- | LDC_Simple_Incomp_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  42.17 s |            --- |         --- | PIP_Simple_Incomp_musubi       |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  58.90 s |            --- |         --- | PIP_Force_Incomp_musubi        |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  36.36 s |            --- |         --- | PIP_ML_Incomp_musubi           |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  95.86 s |            --- |         --- | PIP_LES_Incomp_musubi          |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  29.70 s |            --- |         --- | PIP_Split_Incomp_musubi        |
|  2a225a631e22  |    OK     |   difference    | 0.000%  | 103.56 s |            --- |         --- | ps_cyl2d_musubi                |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |  18.48 s |            --- |         --- | tut_pip_musubi                 |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   3.68 s |            --- |         --- | tut_gauss_pulse                |
|  2a225a631e22  |    OK     |   difference    | 0.000%  |   3.87 s |            --- |         --- | tut_channel_musubi             |
--------------------------------------------------------------------------------------------------------------------------------------
```

Unfortunately the `gaussian2d` setup does *not* work for me. I always end up with `NaN` there.

**Review by haraldkl — COMMENTED** [review]
`gaussian2d` leads to `NaN` in the Musubi simulation. Note that it is possible to use environment variables in Lua, I've changed the `args_given.lua` to illustrate how. But in general generated files should not be part of the repository.

**Review by haraldkl — DISMISSED** [review]
I've adapted the kernel name now, if it is generic for stencil layouts the `_l` should still be there, just not followed by a specific layout. That leaves the gaussian2d setup as an open issue. Other than that it's fine for me to merge this.

**Comment by MikeW097** [comment]
> Is this a replacement for #12 then? Does this mean we should close #12?

Yes, it's now based on the newest version. We can close the previous one.

**Comment by MikeW097** [comment]
> I've adapted the kernel name now, if it is generic for stencil layouts the `_l` should still be there, just not followed by a specific layout. That leaves the gaussian2d setup as an open issue. Other than that it's fine for me to merge this.

Thanks for the change.

The gaussian2d case runs with no problem on my PC. What command did you use to run the case?

**Comment by haraldkl** [comment]
>The gaussian2d case runs with no problem on my PC. What command did you use to run the case?

Sorry for the late reply. I tried to reproduce my problem by running the default `args_given.lua` parameters, but that works for me now. I unfortunately can not remember what I did before my vacation, which might have resulted in the failure I saw.

Should we add the default parameters settings there to the recheck?

File: examples/passive_scalar/benchmark/cylinder2d/args.template [hunk 1]
```diff
@@ -0,0 +1,6 @@
+nelem = 40
+length_usr = 128
+level_usr = math.log(length_usr)/math.log(2)
+t_total = 10
+interval = t_total / 10
+stl_path = '$!stl_path!$'
```
Comment (KannanMasilamani) [inline examples/passive_scalar/benchmark/cylinder2d/args.template:6]: please provide default stl_path and use recheck (shepherd) specific parameter block `$!stl_path!$` only when running this testcase using shepherd. 
For example:
```
shepherd = true
if shepherd then
  stl_path = '$!stl_path!$'
else
  stl_path = './cylinder40.stl'
end
```

See example `examples/fluid/absorbingLayer/acousticCylinder2D/seeder.lua`

File: examples/passive_scalar/benchmark/cylinder2d/timing.res
Comment (KannanMasilamani) [inline examples/passive_scalar/benchmark/cylinder2d/timing.res:1]: We do not add timing.res to the repository. This timing file changes with every run so please remove it.

File: examples/passive_scalar/benchmark/gaussian2d/compare_trt.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+do
+    echo "tau = $tau" > arg_given.lua
+    echo "u_field = 0." >> arg_given.lua
+    echo "order = 'second'" >> arg_given.lua
+    echo "collision = '$collision'" >> arg_given.lua
+    echo "sigma0 = 40" >> arg_given.lua
+
+    musubi
+    lua calculateD.lua >> trt_comparison.res
+done
```
Comment (KannanMasilamani) [inline examples/passive_scalar/benchmark/gaussian2d/compare_trt.sh:15]: Missing indentation!

File: examples/passive_scalar/benchmark/gaussian2d/index.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ is neglected. Results of t=200 are used to calculate the diffusion factor.
+
+![The comparison of concentration between simulation and analytical solution](media/compare_sim_ana_C.jpg)
+![The error of diffusion factor when u\_bg=0](media/err_numerical_diffusion.jpg)
+![Absolute error of D with u\_bg and order of equilibriums](media/d_abs_err_tau0.501.jpg)
+![Absolute error of D with u\_bg and order of equilibriums](media/d_abs_err_tau0.503.jpg)
+![Absolute error of D with u\_bg and order of equilibriums](media/d_abs_err_tau0.8.jpg)
+![Absolute error of D with u\_bg and order of equilibriums](media/d_abs_err_tau2.jpg)
+![Absolute error of D with u\_bg and order of equilibriums](media/d_abs_err_tau5.jpg)
+![The model error for first equilibrium](media/d_1st_model_error.jpg)
```
Comment (KannanMasilamani) [inline examples/passive_scalar/benchmark/gaussian2d/index.md:27]: Please add the script used to create the plots.

File: source/bc/mus_bc_passiveScalar_module.fpp [hunk 4]
```diff
@@ -204,6 +206,110 @@ contains
...
+        if( globBC%elemLvl(iLevel)%bitmask%val(iDir, iElem )) then
+          invDir = layout%fStencil%cxDirInv(iDir)
+
+          state(?FETCH?( iDir, iField, elemPos, QQ, nScalars,nSize, neigh )) = &
+            &   - fTmp( (iElem-1)*nScalars + invDir ) &
+            &   + 2._rk*layout%weight( invDir ) * rhoDef(iElem) 
+
+        end if ! bitMask
+      end do ! iDir
+    end do ! iElem
```
Comment (KannanMasilamani) [inline source/bc/mus_bc_passiveScalar_module.fpp:308]: Just a suggestion. This element loop could be replaced with a link loop see `velocity_bounceback` in `mus_bc_fluid_module.fpp`

File: source/compute/mus_compute_bgk_module.fpp [hunk 5]
```diff
@@ -61,9 +61,11 @@ module mus_bgk_module
 
   private
 
-  public :: mus_advRel_kCFD_rBGK_vStdNoOpt_l
-  public :: bgk_advRel_flekkoy
   public :: bgk_advRel_flekkoy_noFluid
+  public :: mus_advRel_kCFD_rBGK_vStdNoOpt_l
+  public :: mus_advRel_kPS_rBGK_v1st_l
+  public :: mus_advRel_kPS_rBGK_v2nd_l
+  public :: mus_advRel_kPS_rTRT_vStdNoOpt_l
```
Comment (KannanMasilamani) [inline source/compute/mus_compute_bgk_module.fpp:68]: Could you please move all passive scalar routines to seperate module `mus_compute_passiveScalar_module.fpp`

File: examples/passive_scalar/benchmark/cylinder2d/args.template (hunk 1 above)
Comment (haraldkl) [reply examples/passive_scalar/benchmark/cylinder2d/args.template:6]: No need for that here. The `args.template` is only used by shepherd anyway. If run manually we are including `args.lua` instead.

File: source/bc/mus_bc_passiveScalar_module.fpp (hunk 4 above)
Comment (MikeW097) [reply source/bc/mus_bc_passiveScalar_module.fpp:308]: Thanks for the suggestion. But in this case, the "out" population is only computed by its inverse direction of population. So, it's different from "velocity_bounceback" and needs to read from the boundary element in any way. I still keep the original code.

File: examples/passive_scalar/benchmark/cylinder2d/timing.res
Comment (MikeW097) [reply examples/passive_scalar/benchmark/cylinder2d/timing.res:1]: It has been removed.

File: examples/passive_scalar/benchmark/gaussian2d/compare_trt.sh (hunk 2 above)
Comment (MikeW097) [reply examples/passive_scalar/benchmark/gaussian2d/compare_trt.sh]: The file is now removed.

File: examples/passive_scalar/benchmark/gaussian2d/index.md (hunk 3 above)
Comment (MikeW097) [reply examples/passive_scalar/benchmark/gaussian2d/index.md]: The plot scripts have been added.

File: source/compute/mus_compute_bgk_module.fpp (hunk 5 above)
Comment (MikeW097) [reply source/compute/mus_compute_bgk_module.fpp]: The code has been moved to a new file named `mus_compute_passiveScalar_module.fpp`.

**Review by KannanMasilamani — APPROVED** [review]
Thank you for your changes. Everything looks fine. 
Just a minor comment, please use just two indent spaces in lua files. Some files have four spaces. It is not a big issue but it would be nice to have a consistent spacing.

**Comment by haraldkl** [comment]
@MikeW097 Thanks a lot. It's merged and available in the parent musubi repository with commit [530eeec9](https://github.com/apes-suite/musubi/commit/530eeec95e089bd78b0fc4a6012ac0cf777ef105).

**Comment by MikeW097** [comment]
> Thank you for your changes. Everything looks fine. Just a minor comment, please use just two indent spaces in lua files. Some files have four spaces. It is not a big issue but it would be nice to have a consistent spacing.

Thanks for the review and pointing out the issue. I will take more care of the spacing.

**Comment by MikeW097** [comment]
> @MikeW097 Thanks a lot. It's merged and available in the parent musubi repository with commit [530eeec9](https://github.com/apes-suite/musubi/commit/530eeec95e089bd78b0fc4a6012ac0cf777ef105).

Thanks a lot, Harald!

**Pull request merged**